File: src/model/instrument/id.zig
const std = @import("std");
const errors = @import("../../core/errors.zig");

pub const Id = union(enum) {
    unset,
    generic,
    tropomi,
    synthetic,
    custom: []const u8,

    pub fn parse(value: []const u8) Id {
        if (value.len == 0) return .unset;
        if (std.mem.eql(u8, value, "generic")) return .generic;
        if (std.mem.eql(u8, value, "tropomi")) return .tropomi;
        if (std.mem.eql(u8, value, "synthetic")) return .synthetic;
        return .{ .custom = value };
    }

    pub fn label(self: Id) []const u8 {
        return switch (self) {
            .unset => "",
            .generic => "generic",
            .tropomi => "tropomi",
            .synthetic => "synthetic",
            .custom => |value| value,
        };
    }

    pub fn validate(self: Id) errors.Error!void {
        switch (self) {
            .unset => return errors.Error.MissingObservationInstrument,
            .custom => |value| if (value.len == 0) return errors.Error.MissingObservationInstrument,
            .generic, .tropomi, .synthetic => {},
        }
    }
};

File: rust_src/input/instrument/id.rs
use crate::common::errors;

#[derive(Debug, Default, Clone, PartialEq, Eq)]
pub enum Id {
    #[default]
    Unset,
    Generic,
    Tropomi,
    Synthetic,
    Custom(String),
}

impl Id {
    pub fn parse(value: &str) -> Self {
        match value {
            "" => Self::Unset,
            "generic" => Self::Generic,
            "tropomi" => Self::Tropomi,
            "synthetic" => Self::Synthetic,
            _ => Self::Custom(value.to_string()),
        }
    }

    pub fn label(&self) -> &str {
        match self {
            Self::Unset => "",
            Self::Generic => "generic",
            Self::Tropomi => "tropomi",
            Self::Synthetic => "synthetic",
            Self::Custom(value) => value,
        }
    }

    pub fn validate(&self) -> Result<(), errors::Error> {
        match self {
            Self::Unset => Err(errors::Error::MissingObservationInstrument),
            Self::Custom(value) if value.is_empty() => {
                Err(errors::Error::MissingObservationInstrument)
            }
            Self::Generic | Self::Tropomi | Self::Synthetic | Self::Custom(_) => Ok(()),
        }
    }
}
